File: tests/bench/09_stdlib_math.go
// Benchmark 09: Stdlib Math — trigonometry, sqrt, pow, log
// Measures math function call overhead

package main

import (
	"fmt"
	"math"
	"time"
)

func main() {
	// Part A: sqrt loop
	nSqrt := 10_000_000
	start := time.Now()
	total := 0.0
	for i := 1; i <= nSqrt; i++ {
		total += math.Sqrt(float64(i))
	}
	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("[bench] math_sqrt: %d calls in %d ms (result=%d)\n", nSqrt, elapsed, int64(math.Floor(total)))

	// Part B: sin + cos
	nTrig := 5_000_000
	start = time.Now()
	sum := 0.0
	for i := 0; i < nTrig; i++ {
		x := float64(i) * 0.001
		sum += math.Sin(x) + math.Cos(x)
	}
	elapsed = time.Since(start).Milliseconds()
	fmt.Printf("[bench] math_trig: %d sin+cos pairs in %d ms (result=%d)\n", nTrig, elapsed, int64(math.Floor(sum)))

	// Part C: pow + log roundtrip
	nPowlog := 5_000_000
	start = time.Now()
	acc := 0.0
	for i := 1; i <= nPowlog; i++ {
		acc += math.Log(math.Pow(2.0, float64(i%20)) + 1.0)
	}
	elapsed = time.Since(start).Milliseconds()
	fmt.Printf("[bench] math_powlog: %d pow+log calls in %d ms (result=%d)\n", nPowlog, elapsed, int64(math.Floor(acc)))
}

File: tests/bench/09_stdlib_math.rs
// Benchmark 09: Stdlib Math — trigonometry, sqrt, pow, log
// Measures math function call overhead

use std::time::Instant;

fn main() {
    // Part A: sqrt loop
    let n_sqrt: i64 = 10_000_000;
    let start = Instant::now();
    let mut total: f64 = 0.0;
    for i in 1..=n_sqrt {
        total += (i as f64).sqrt();
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] math_sqrt: {} calls in {} ms (result={})", n_sqrt, elapsed, total.floor() as i64);

    // Part B: sin + cos
    let n_trig: i64 = 5_000_000;
    let start = Instant::now();
    let mut sum: f64 = 0.0;
    for i in 0..n_trig {
        let x = i as f64 * 0.001;
        sum += x.sin() + x.cos();
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] math_trig: {} sin+cos pairs in {} ms (result={})", n_trig, elapsed, sum.floor() as i64);

    // Part C: pow + log roundtrip
    let n_powlog: i64 = 5_000_000;
    let start = Instant::now();
    let mut acc: f64 = 0.0;
    for i in 1..=n_powlog {
        acc += ((2.0_f64).powf((i % 20) as f64) + 1.0).ln();
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] math_powlog: {} pow+log calls in {} ms (result={})", n_powlog, elapsed, acc.floor() as i64);
}
